server/etcdserver/raft.go

```diff
@@ -237,6 +244,12 @@ func (r *raftNode) start(rh *raftReadyHandler) {
 				if err := r.storage.Save(rd.HardState, rd.Entries); err != nil {
 					r.lg.Fatal("failed to save Raft hard state and entries", zap.Error(err))
 				}
+
+				if walNotifyc != nil {
```

Comment:
I think you can remove snapNotifyc and walNotifyc and just use `r.applyc <- ap` here for 1 node cluster (instead of line 220).
What does snapNotifyc achieve? If I understand this correctly it will allow applyAll->applySnapshot to run sooner. But this is a 1 node cluster case, so applySnapshot from leader will never be triggered because there are no followers.

If my logic is correct, then you can also rename shouldWaitWALSync() to isSingleNodeCluster()

Reply:
It isn't good to mix the new patch/fix with the existing code/logic, because it may break the existing functionality and cause regression. 

Note that there may be multiple entries (including both unstable and committed entries for single-server cluster) in `Ready`, and it may also have zero entry for raft `no-op`. The current implementation triggers the responding to client when finish applying each entry, but we can't send&receive the `walNotifyc` multiple times.

The above explanation may look complicated. The point here is it's really subtle and error prone, and the best practice for this case is to create a separate channel for this requirement (waiting for WAL sync).